:sparkles: add WETH intents for op, main, poly, arb and base

Quazia opened the pull request:
closes #3 

I noticed theres tests for the RPC and for the client, but there isn't a case of testing against chain state. I can write unit tests for this but I would want them to run against a forked instance of chain state vs mocking out the on-chain interactions. Let me know if there's appetite for that work and I can either do it in a separate PR or in this one (I tend to think it'd make sense to build out in a seperate PR and then backfill unit tests for the tools).

Also followed the general pattern for the transfer tool - let me know if there are any requested changes and appreciate your patience.

nerderlyne commented on src/shared/weth/index.ts:
```diff
@@ -0,0 +1,6 @@
+import {depositWETHIntent, withdrawWETHIntent} from "./intents";
+
+export function wethTools(): BaseTool[] {
+  return createToolCollection([depositWETHIntent, withdrawWETHIntent]);
```
update to intent_depositWETH and intent_withdrawWETHIntent

nerderlyne commented on src/shared/weth/intents.ts:
```diff
@@ -0,0 +1,119 @@
+const depositWETHChains = supportedChains;
+const withdrawWETHChains = supportedChains;
+
+export const depositWETHIntent = createTool({
```
update to intent_depositWETH

nerderlyne commented on src/shared/weth/intents.ts:
```diff
@@ -0,0 +1,119 @@
+
+const depositWETHParameters = z.object({
+  chainId: z.number().describe("Chain ID to deposit on"),
+  amount: z.string().describe("Amount of ETH to deposit (as a string)"),
```
should this be a number (ether) ?

nerderlyne commented on src/shared/weth/intents.ts:
```diff
@@ -0,0 +1,119 @@
+
+const withdrawWETHParameters = z.object({
+  chainId: z.number().describe("Chain ID to withdraw on"),
+  amount: z.string().describe("Amount of WETH to withdraw (as a string)"),
```
should this be a number (ether) ?

nerderlyne commented on src/shared/weth/intents.ts:
```diff
@@ -0,0 +1,119 @@
+
+    if (!walletClient) {
+      return {
+        intent: `Deposit ${amount} ETH into WETH on chain ${chainId}`,
```
`Deposit ${amount} ETH into WETH`

chainId is returned in obj

nerderlyne commented on src/shared/weth/intents.ts:
```diff
@@ -0,0 +1,119 @@
+        } as SendTransactionParameters);
+    
+    return {
+      intent: `Deposit ${amount} ETH into WETH on chain ${chainId}`,
```
Deposit ${amount} ETH into WETH

chainId is returned in obj

nerderlyne commented on src/shared/weth/intents.ts:
```diff
@@ -0,0 +1,119 @@
+
+    if (!walletClient) {
+      return {
+        intent: `Withdraw ${amount} WETH to native ETH on chain ${chainId}`,
```
`Unwrap ${amount} WETH to ETH`

chainId is returned in obj

nerderlyne commented on src/shared/weth/intents.ts:
```diff
@@ -0,0 +1,119 @@
+      } as SendTransactionParameters);
+
+      return {
+        intent: `Withdraw ${amount} WETH to native ETH on chain ${chainId}`,
```
`Unwrap ${amount} WETH to ETH`

chainId is returned in obj

nerderlyne commented on src/shared/weth/constants.ts:
```diff
@@ -0,0 +1,32 @@
+
+export const supportedChains = [mainnet, optimism, arbitrum, polygon];
+
+export const WETH_ADDRESS = {
```
Add base support

nerderlyne reviewed (changes requested):
Thank you so much for doing this! It looks great. There are some minor changes - we are following `intent` prefix syntax to make it clear the tool is a transaction intent for consumers/statically(I will add this to contribution guidelines). Also support for base will be great. This is awesome though.

nerderlyne commented:
Separate PR is good idea. We can have unit tests for tools itself against the forked state (using anvil perhaps). We want to also add more thorough integrated tests of the agent itself (including planning, task decomposition and so on) but let's start with the unit tests for the tools itself. Would appreciate you taking that up!

Thanks for your contribution on this PR!

Quazia commented:
@nerderlyne I would also default to Anvil for testing forked state, we could use HH but preference for _not_ doing that lol. There are some nice qualities in terms of parallelization of tests and resets but the cons outweigh the pros in my experience.


All review comments should be address, I fixed the prefix for the intent functions, added base support, and switched the typing for the amount. I'm fairly neutral on number vs string for amount but I separated out the PRs to provide optionality on the fixes. Thanks for you review and the warm welcome to the repo 😊

nerderlyne commented:
I think llm is likely respond in ether so should convert to wei using `parseEther` in both functions and have explicit instructions regarding format in the zod schema. Currently assumption seems wei

Quazia commented:
Ok - idk enough about how the model interacts with the params to know if it makes sense to type the param as anumber and then convert it to a string to pass to the parse call but I implemented it that way. lmk if this looks right.

nerderlyne commented:
<img width="1068" alt="Screenshot 2025-01-30 at 4 40 16 PM" src="https://github.com/user-attachments/assets/75022b50-a31c-45fc-92d9-61f5948953d4" />
- Changed zod schema to accept number (ether) instead of bigint, as you can see bigint causes an error. We don't accept strings because we want to reject responses like "ten" when parsing. 
- Also added waiting for receipt so agent can wrap and unwrap within the same call and not get an error due to prev. transaction not being accepted on the chain yet(will implement this in other tools too!) 

<img width="1068" alt="Screenshot 2025-01-30 at 4 46 05 PM" src="https://github.com/user-attachments/assets/de3b0e43-feed-4ab5-a77b-3f51208136d8" />

It works perfectly now I added sepolia so we can just run basic examples there 

https://sepolia.etherscan.io/tx/0x6677076e584ab279efb35031806c8c8942162437fde8ee2392e65d9f157432e8 

https://sepolia.etherscan.io/tx/0x3b0f2a33ff4ffd99f886ed203eb8ef2b0cc8af818d83768418627748ef125d71

Thank you again for working on this. Merging 🩵
